### ISSUE TITLE
prefetch_to_device doesn't overlap copy(HtoD) with computation and also may fail

### ISSUE THREAD

**Issue by sh1ng** [tensorflow/tensorflow#20628]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Ubuntu 16.04.4 LTS
- **TensorFlow installed from (source or binary)**: pip3
- **TensorFlow version (use command below)**:1.8.0
- **Python version**: 3
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:9.0
- **GPU model and memory**:GTX 1070
- **Exact command to reproduce**:
```
import tensorflow as tf
import numpy as np

x = tf.placeholder(shape=[None, 128], dtype=np.float32)
y = tf.placeholder(shape=[None], dtype=np.float32)

ds = tf.data.Dataset.from_tensor_slices((x, y))
ds = ds.repeat()
ds = ds.batch(1024)
############################################
prefetch = 1024 # if it's bigger than 512 it results in an error
############################################
ds = ds.apply(tf.contrib.data.prefetch_to_device("/gpu:0", prefetch))
it = ds.make_initializable_iterator()

w = tf.get_variable(
                    name='w',
                    initializer=tf.contrib.layers.xavier_initializer(),
                    shape=[128, 1],
                    dtype=tf.float32)
b = tf.get_variable(
                    name='b',
                    initializer=tf.contrib.layers.xavier_initializer(),
                    shape=[1],
                    dtype=tf.float32)

optimizer = tf.train.AdamOptimizer(1e-3)

next_x, next_y = it.get_next()
prediction = tf.matmul(next_x, w) + b
train = optimizer.minimize((prediction - next_y)**2)

with tf.Session() as session:
    x_val = np.random.normal(size=(1024*1024, 128)).astype(np.float32)
    y_val = np.random.normal(size=(1024 * 1024)).astype(np.float32)
    session.run(tf.global_variables_initializer())
    session.run(it.initializer, feed_dict={x: x_val, y: y_val})
    for _ in range(1024):
        session.run(train)
```

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
v1.8.0-0-g93bc2e2072 1.8.0

### Describe the problem

If I specify `prefetch` size to 128 I don't see any parallel data copy(HtoD) overlapping gpu computation(check screenshot and nvidia profiler report).
<img width="1072" alt="screen shot 2018-07-08 at 17 36 25" src="https://user-images.githubusercontent.com/2821871/42421527-412a3124-82d7-11e8-94e7-1f9114d88cec.png">

And If I increase it it fails
```
$ python3 test.py 
2018-07-08 17:32:18.472300: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-07-08 17:32:18.550265: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:898] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-07-08 17:32:18.550841: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1356] Found device 0 with properties: 
name: GeForce GTX 1070 major: 6 minor: 1 memoryClockRate(GHz): 1.683
pciBusID: 0000:01:00.0
totalMemory: 7.93GiB freeMemory: 7.83GiB
2018-07-08 17:32:18.550852: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1435] Adding visible gpu devices: 0
2018-07-08 17:32:18.711860: I tensorflow/core/common_runtime/gpu/gpu_device.cc:923] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-08 17:32:18.711888: I tensorflow/core/common_runtime/gpu/gpu_device.cc:929]      0 
2018-07-08 17:32:18.711907: I tensorflow/core/common_runtime/gpu/gpu_device.cc:942] 0:   N 
2018-07-08 17:32:18.712052: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1053] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7568 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1070, pci bus id: 0000:01:00.0, compute capability: 6.1)
2018-07-08 17:32:25.310286: E tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:650] failed to record completion event; therefore, failed to create inter-stream dependency
2018-07-08 17:32:25.310311: I tensorflow/stream_executor/stream.cc:4737] stream 0x1448efe0 did not memcpy host-to-device; source: 0x7f8feae97000
2018-07-08 17:32:25.310316: E tensorflow/stream_executor/stream.cc:309] Error recording event in stream: error recording CUDA event on stream 0x1448f080: CUDA_ERROR_DEINITIALIZED; not marking stream as bad, as the Event object may be at fault. Monitor for further errors.
2018-07-08 17:32:25.310323: E tensorflow/stream_executor/cuda/cuda_event.cc:49] Error polling for event status: failed to query event: CUDA_ERROR_DEINITIALIZED
2018-07-08 17:32:25.310328: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:208] Unexpected Event status: 1
Aborted (core dumped)
```

It'd be nice to have some canonical examples how to use it.

[prefetch_on_device.nvvp.zip](https://github.com/tensorflow/tensorflow/files/2173870/prefetch_on_device.nvvp.zip)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sh1ng** [comment]
@rohan100jain any progress?

**Comment by rohan100jain** [comment]
So usually I'd recommend a prefetch buffer size of something in the range of 1-4. 1 should mostly work because you want to pipeline the next computation while the current one is going on. With this large number, you're just creating a lot of memory pressure on the GPU without much benefit. 

Also, can you increase the size of the data that is being transferred over? Make it some really large tensor. Then we might be able to see the overlap. ITs possible that overlap is possible even in this scenario but the copied tensor is small and is competing with other CPU work (not shown on this GPU profile).

**Comment by sh1ng** [comment]
Ok, I've changed the script a bit. BTW now it's TF 1.10
Briefly I set `prefetch_to_device` buffer to 1(it could be described in the documentation that it's multiplier for batch size) and input size from 128 to 1024 to make computation more time consuming. 
```
import tensorflow as tf
import numpy as np

x = tf.placeholder(shape=[None, 1024], dtype=np.float32)
y = tf.placeholder(shape=[None], dtype=np.float32)

ds = tf.data.Dataset.from_tensor_slices((x, y))
ds = ds.repeat()
ds = ds.batch(1024)
ds = ds.apply(tf.contrib.data.prefetch_to_device("/gpu:0", 1))
it = ds.make_initializable_iterator()

w = tf.get_variable(
                    name='w',
                    initializer=tf.contrib.layers.xavier_initializer(),
                    shape=[1024, 1],
                    dtype=tf.float32)
b = tf.get_variable(
                    name='b',
                    initializer=tf.contrib.layers.xavier_initializer(),
                    shape=[1],
                    dtype=tf.float32)

optimizer = tf.train.AdamOptimizer(1e-3)

next_x, next_y = it.get_next()
prediction = tf.matmul(next_x, w) + b
train = optimizer.minimize((prediction - next_y)**2)

gpu_options = tf.GPUOptions(force_gpu_compatible=True)
config = tf.ConfigProto(gpu_options=gpu_options, allow_soft_placement=True)
with tf.Session() as session:
    x_val = np.random.normal(size=(1024*1024, 1024)).astype(np.float32)
    y_val = np.random.normal(size=(1024 * 1024)).astype(np.float32)
    session.run(tf.global_variables_initializer())
    session.run(it.initializer, feed_dict={x: x_val, y: y_val})
    for _ in range(1024):
        session.run(train)


```

I don't see any improvements at all. 
Also there're a big gap after each iteration(it's longer than H2D copy plus computation itself). Is it a TF or/and python overhead?

I've implemented before overlapping copying into a device with computation in c++. And result was predictable and much faster. Does make sense to use TF C++ API? 

A new screenshot and profiling report are attached. 
![screen shot 2018-08-17 at 12 29 14](https://user-images.githubusercontent.com/2821871/44262247-a66fe300-a21a-11e8-9f3d-18ed16ed957f.png)
[prefetch_on_device.nvvp 2.zip](https://github.com/tensorflow/tensorflow/files/2296907/prefetch_on_device.nvvp.2.zip)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jiazhe0909** [comment]
Same issue here, a ctx synchronize always block the H2D memory before the complete of GPU computing. @sh1ng I saw a lot of space between two H2D in your profiling results, maybe you should reduce the cpu processing time first.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sh1ng** [comment]
@jiazhe0909 I've posted simple example and there's no processing at all, besides 
```
ds = tf.data.Dataset.from_tensor_slices((x, y))
ds = ds.repeat()
ds = ds.batch(1024)
```
If it causes intensive CPU processing, there's an issue in `tf.data.Dataset`, but I'd like to get some feedback from the team.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by billythegod** [comment]
Same issue with tf 2.1.0, using model.fit and tf.keras, no overlap even when using copy_to_device.

**Comment by papillonyi** [comment]
Same issue with tf 2.4.0, no overlap when using copy_to_device or prefetch_to_device

**Comment by jsimsa** [comment]
Please see https://github.com/tensorflow/tensorflow/issues/43905#issuecomment-823675760

**Comment by Saduf2019** [comment]
@sh1ng 
Could you please try on latest stable version of tf and let us know if this is still an issue.Thanks!
